### PR TITLE
Bugfix MTE-402 [v111] Add more checks to ensure pop-up disappears

### PR DESCRIPTION
**Pull request by clarmso**
`HistoryTests/testClearPrivateDataButtonDisabled()` fails intermittently. The test expects the "Clear Private Data" button to be disabled, but in practice the button may still be enabled after the private data has been cleared.

I observed that the "Clear Private Data" button may not be disabled right away when the alert is closed. The button eventually becomes disabled. Let me add a `sleep()` to allow enough time for the button to transition to disabled.

I can't reproduce the failure on my end using `iPhone 14` or `iPad Pro (12.9-inch) (5th generation)` even after the test has been run 100 times.

https://mozilla-hub.atlassian.net/browse/MTE-402

### CODE REVIEW

**Review by isabelrios — APPROVED** [review]
I would have preferred not to add a sleep in a test that runs as part of the smoketest as this will imply that delay in all PRs... but if there is not any other way for now, let's add it and monitor the test closely

**Comment by clarmso** [comment]
> I would have preferred not to add a sleep in a test that runs as part of the smoketest as this will imply that delay in all PRs... but if there is not any other way for now, let's add it and monitor the test closely

Same here. I'd rather not to add a `sleep()` as well. I haven't found a way to wait until the "Clear Private Data" button is disabled. Let me find other elements on the page to wait for.

**Review by isabelrios — APPROVED** [review]

**Comment by isabelrios** [comment]
@clarmso the test failed although it passes in a retry.. looking at the error, my feeling is that is more related to not finding the Menu to open Settings from a Tab: `"Failed to scroll to visible (by AX action) Button, {{312.0, 764.0}, {78.0, 46.0}}, identifier: '**TabToolbar.menuButton**'` So, in addition to the fix in this PR, we should investigate how to fix that before landing this PR which is only about this test.

**Comment by clarmso** [comment]
> @clarmso the test failed although it passes in a retry.. looking at the error, my feeling is that is more related to not finding the Menu to open Settings from a Tab: `"Failed to scroll to visible (by AX action) Button, {{312.0, 764.0}, {78.0, 46.0}}, identifier: '**TabToolbar.menuButton**'` So, in addition to the fix in this PR, we should investigate how to fix that before landing this PR which is only about this test.

🤔 💭 The `TabToolbar.menuButton` is the hamburger menu under the `Navigation Toolbar`. The `app.debugDescription` looks like the following:
```
                    Other, 0x7f91cca11570, {{0.0, 772.0}, {393.0, 80.0}}, label: 'Navigation Toolbar'
                      Other, 0x7f91cca11680, {{0.0, 772.0}, {393.0, 46.0}}
                        Button, 0x7f91cca11790, {{0.0, 772.0}, {78.7, 46.0}}, identifier: 'TabToolbar.backButton', label: 'Back', Disabled
                        Button, 0x7f91cca118a0, {{78.7, 772.0}, {78.7, 46.0}}, identifier: 'TabToolbar.forwardButton', label: 'Forward', Disabled
                        Button, 0x7f91cca119b0, {{157.3, 772.0}, {78.3, 46.0}}, identifier: 'TabToolbar.homeButton', label: 'Search'
                        Button, 0x7f91cca11ac0, {{235.7, 772.0}, {78.7, 46.0}}, identifier: 'TabToolbar.tabsButton', label: 'Show Tabs', value: 1
                          StaticText, 0x7f91cca11bd0, {{263.0, 783.0}, {24.0, 24.0}}, label: '1'
                        Button, 0x7f91cca11ce0, {{314.3, 772.0}, {78.7, 46.0}}, identifier: 'TabToolbar.menuButton', label: 'Menu'
```
A reason could be caused by the keyboard not closing up. Before the keyboard is closed, `TabToolbar.menuButton` is present in `app.debugDescription` but is not necessarily clickable or visible.
Screenshot 2023-01-31 at 3 40 39 PM

**Comment by clarmso** [comment]
I just encounter such a failure on the `menuButton` locally just now. (Sorry, no video.) Here's the logs:
```
2023-01-31 16:20:38.265312-0500 XCUITests-Runner[47410:551258] Running tests...
Test Suite 'Selected tests' started at 2023-01-31 16:20:40.233
Test Suite 'XCUITests.xctest' started at 2023-01-31 16:20:40.235
Test Suite 'HistoryTests' started at 2023-01-31 16:20:40.235
Test Case '-[XCUITests.HistoryTests testClearPrivateDataButtonDisabled]' started.
    t =     0.00s Start Test at 2023-01-31 16:20:40.238
    t =     0.06s Set Up
    t =     0.08s     Open org.mozilla.ios.Fennec
    t =     0.10s         Launch org.mozilla.ios.Fennec
    t =     2.96s             Setting up automation session
    t =     4.70s             Wait for org.mozilla.ios.Fennec to idle
    t =    10.42s Checking `Expect predicate `exists == 1` for object "urlBar-cancel" Button`
    t =    10.42s     Checking existence of `"urlBar-cancel" Button`
    t =    11.71s Checking `Expect predicate `exists == 1` for object "TabToolbar.menuButton" Button`
    t =    11.71s     Checking existence of `"TabToolbar.menuButton" Button`
    t =    12.83s Checking `Expect predicate `exists == 1` for object "TabToolbar.menuButton" Button`
    t =    12.83s     Checking existence of `"TabToolbar.menuButton" Button`
    t =    12.92s Tap "TabToolbar.menuButton" Button
    t =    12.92s     Wait for org.mozilla.ios.Fennec to idle
    t =    12.94s     Find the "TabToolbar.menuButton" Button
    t =    13.02s     Check for interrupting elements affecting "TabToolbar.menuButton" Button
    t =    13.09s     Synthesize event
    t =    13.13s         Scroll element to visible
    t =    13.13s         Find the "TabToolbar.menuButton" Button
    t =    13.22s         Computed hit point {-1, -1} after scrolling to visible
    t =    13.22s         Failed: Failed to scroll to visible (by AX action) Button, {{312.0, 764.0}, {78.0, 46.0}}, identifier: 'TabToolbar.menuButton', label: 'Menu', error: Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103
    t =    14.22s     Retrying `Tap "TabToolbar.menuButton" Button` (attempt #2)
    t =    14.22s         Wait for org.mozilla.ios.Fennec to idle
    t =    14.26s         Find the "TabToolbar.menuButton" Button
    t =    14.34s         Check for interrupting elements affecting "TabToolbar.menuButton" Button
    t =    14.40s         Synthesize event
    t =    14.43s             Scroll element to visible
    t =    14.43s             Find the "TabToolbar.menuButton" Button
    t =    14.52s             Computed hit point {-1, -1} after scrolling to visible
    t =    14.52s             Failed: Failed to scroll to visible (by AX action) Button, {{312.0, 764.0}, {78.0, 46.0}}, identifier: 'TabToolbar.menuButton', label: 'Menu', error: Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103
    t =    15.53s     Retrying `Tap "TabToolbar.menuButton" Button` (attempt #3)
    t =    15.53s         Wait for org.mozilla.ios.Fennec to idle
    t =    15.55s         Find the "TabToolbar.menuButton" Button
    t =    15.63s         Check for interrupting elements affecting "TabToolbar.menuButton" Button
    t =    15.70s         Synthesize event
    t =    15.73s             Scroll element to visible
    t =    15.73s             Find the "TabToolbar.menuButton" Button
    t =    15.82s             Computed hit point {-1, -1} after scrolling to visible
    t =    15.82s             Failed: Failed to scroll to visible (by AX action) Button, {{312.0, 764.0}, {78.0, 46.0}}, identifier: 'TabToolbar.menuButton', label: 'Menu', error: Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103
/Users/cso/Library/Developer/Xcode/DerivedData/Client-enucdobedxsoydgzdifsvuwxgrkz/SourcePackages/checkouts/MappaMundi/Sources/MMScreenStateNode.swift:92: error: -[XCUITests.HistoryTests testClearPrivateDataButtonDisabled] : Failed to Error Domain=com.apple.dt.xctest.ui-testing.error Code=10202 "Failed to scroll to visible (by AX action) Button, {{312.0, 764.0}, {78.0, 46.0}}, identifier: 'TabToolbar.menuButton', label: 'Menu', error: Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103" UserInfo={NSLocalizedDescription=Failed to scroll to visible (by AX action) Button, {{312.0, 764.0}, {78.0, 46.0}}, identifier: 'TabToolbar.menuButton', label: 'Menu', error: Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103, NSUnderlyingError=0x600003675e60 {Error Domain=XCTDaemonErrorDomain Code=0 "Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103" UserInfo={NSLocalizedDescription=Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103}}}: Failed to scroll to visible (by AX action) Button, {{312.0, 764.0}, {78.0, 46.0}}, identifier: 'TabToolbar.menuButton', label: 'Menu', error: Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103. (Underlying Error: Error kAXErrorCannotComplete performing AXAction kAXScrollToVisibleAction on element AX element pid: 47438, elementOrHash.elementID: 140169612541696.103)
    t =    15.94s Tear Down
    t =    15.94s     Terminate org.mozilla.ios.Fennec:47438
Test Case '-[XCUITests.HistoryTests testClearPrivateDataButtonDisabled]' failed (17.045 seconds).
Test Suite 'HistoryTests' failed at 2023-01-31 16:20:57.282.
         Executed 1 test, with 1 failure (0 unexpected) in 17.045 (17.047) seconds
Test Suite 'XCUITests.xctest' failed at 2023-01-31 16:20:57.283.
         Executed 1 test, with 1 failure (0 unexpected) in 17.045 (17.048) seconds
Test Suite 'Selected tests' failed at 2023-01-31 16:20:57.284.
         Executed 1 test, with 1 failure (0 unexpected) in 17.045 (17.051) seconds
2023-01-31 16:21:05.761 xcodebuild[47201:549100] [MT] IDETestOperationsObserverDebug: 62.106 elapsed -- Testing started completed.
2023-01-31 16:21:05.761 xcodebuild[47201:549100] [MT] IDETestOperationsObserverDebug: 0.000 sec, +0.000 sec -- start
2023-01-31 16:21:05.761 xcodebuild[47201:549100] [MT] IDETestOperationsObserverDebug: 62.106 sec, +62.106 sec -- end
```
🤷🏼‍♀️ I have no idea why `urlBar-cancel` is not clicked during `Action.CloseURLBarOpen`.

File: Tests/XCUITests/HistoryTests.swift [hunk 1]
```diff
@@ -91,6 +91,7 @@ class HistoryTests: BaseTestCase {
             waitForExistence(app.buttons["urlBar-cancel"], timeout: TIMEOUT_LONG)
         }
         navigator.performAction(Action.CloseURLBarOpen)
+        waitForNoExistence(app.buttons["urlBar-cancel"], timeoutValue: TIMEOUT)
```
Comment (clarmso) [inline Tests/XCUITests/HistoryTests.swift:94]: Let's ensure that the keyboard is closed before clicking the `settingsMenuButton`.

Note that when the keyboard is displayed, the `settingsMenuButton` is deemed exist and is hittable, but cannot be tapped. 🤷🏼‍♀️ Let's put a "guard" here before the test attempt to test the presence and tap on `settingsMenuButton.

**Comment by isabelrios** [comment]
I think the probelm with this test is that this is the first one to run as part of the Smoketest3, the first one of the History test suite. What I have noticed after downloading bitrise screenshots is that the test does not even start... it fails to tap on that button and not in the close url bar because we have that first step wrapped into a [XCTExpectedFailure](https://github.com/mozilla-mobile/firefox-ios/blob/478902ac383192bcb8b9e2e77098b0d43b8627c8/Tests/XCUITests/HistoryTests.swift#L90) and so we fail in l95 instead.

I think we can close this PR and focus on finding a way to be sure the app is launched before the tests start...

**Comment by clarmso** [comment]
> I think we can close this PR and focus on finding a way to be sure the app is launched before the tests start...

Would the solution be in the `BaseTestCase` class instead? After `app.launch()` is called, there's no checking whether the app is ready to be tested.

**Comment by isabelrios** [comment]
> > I think we can close this PR and focus on finding a way to be sure the app is launched before the tests start...
> 
> Would the solution be in the `BaseTestCase` class instead? After `app.launch()` is called, there's no checking whether the app is ready to be tested.

Not sure where, to be honest... :/ @jjSDET had some initial thoughts about that IIRC

**Comment by clarmso** [comment]
This work may be superseded by the MappaMundi update.

**Pull request closed without merging**